File: packages/admin/src/facades/MemberInfoFacade.ts
import { type Result, Ok, Err, DomainError } from '@ltdjms/shared';
import {
  BalanceService,
  GameTokenService,
  CurrencyTransactionService,
  GameTokenTransactionService,
  type BalanceView,
  type CurrencyTransaction,
  type GameTokenTransaction,
  type TransactionPage,
} from '@ltdjms/economy';
import {
  RedemptionService,
  type RedemptionResult,
} from '@ltdjms/shop';

/**
 * Summary view combining balance and token info for the user panel.
 */
export interface MemberPanelView {
  readonly guildId: string;
  readonly userId: string;
  readonly balance: number;
  readonly currencyName: string;
  readonly currencyIcon: string;
  readonly tokens: number;
}

/**
 * A single redemption transaction entry in the user panel.
 */
export interface RedemptionTransactionEntry {
  readonly id: number;
  readonly productName: string;
  readonly code: string;
  readonly rewardedAmount: number | null;
  readonly createdAt: Date;
}

/**
 * Paginated redemption transaction history.
 */
export interface RedemptionTransactionPage {
  readonly items: RedemptionTransactionEntry[];
  readonly hasNext: boolean;
  readonly totalPages: number;
  readonly currentPage: number;
}

/**
 * Facade for member-facing queries.
 * Aggregates BalanceService, GameTokenService, transaction services, and redemption.
 * Matches Java MemberInfoFacade.
 */
export class MemberInfoFacade {
  constructor(
    private readonly balanceService: BalanceService,
    private readonly tokenService: GameTokenService,
    private readonly currencyTxService: CurrencyTransactionService,
    private readonly tokenTxService: GameTokenTransactionService,
    private readonly redemptionService: RedemptionService,
  ) {}

  /**
   * Gets a combined view of the member's balance and token info.
   *
   * @see getMemberSummary — alias for the same method
   */
  async getUserPanelView(
    guildId: string,
    userId: string,
  ): Promise<Result<MemberPanelView, DomainError>> {
    return this.getMemberSummary(guildId, userId);
  }

  /**
   * Alias for {@link getUserPanelView}.
   * Provides a consistent naming convention aligned with MemberPanelView.
   */
  async getMemberSummary(
    guildId: string,
    userId: string,
  ): Promise<Result<MemberPanelView, DomainError>> {
    try {
      const balanceView = await this.balanceService.getBalance(Number(guildId), Number(userId));
      const tokenBalance = await this.tokenService.getBalance(Number(guildId), Number(userId));

      return new Ok({
        guildId,
        userId,
        balance: balanceView.balance,
        currencyName: balanceView.currencyName,
        currencyIcon: balanceView.currencyIcon,
        tokens: tokenBalance,
      });
    } catch (err) {
      return new Err(
        DomainError.persistenceFailure(
          `Failed to get user panel view for guildId=${guildId}, userId=${userId}`,
          err instanceof Error ? err : undefined,
        ),
      );
    }
  }

  /**
   * Gets a paginated list of currency transactions for a member.
   */
  async getCurrencyTransactionPage(
    guildId: string,
    userId: string,
    page: number = 1,
    pageSize: number = 10,
  ): Promise<Result<TransactionPage<CurrencyTransaction>, DomainError>> {
    try {
      const txPage = await this.currencyTxService.getTransactionPage(
        Number(guildId),
        Number(userId),
        page,
        pageSize,
      );
      return new Ok(txPage);
    } catch (err) {
      return new Err(
        DomainError.persistenceFailure(
          `Failed to get currency transactions for guildId=${guildId}, userId=${userId}`,
          err instanceof Error ? err : undefined,
        ),
      );
    }
  }

  /**
   * Gets a paginated list of token transactions for a member.
   */
  async getTokenTransactionPage(
    guildId: string,
    userId: string,
    page: number = 1,
    pageSize: number = 10,
  ): Promise<Result<TransactionPage<GameTokenTransaction>, DomainError>> {
    try {
      const txPage = await this.tokenTxService.getTransactionPage(
        Number(guildId),
        Number(userId),
        page,
        pageSize,
      );
      return new Ok(txPage);
    } catch (err) {
      return new Err(
        DomainError.persistenceFailure(
          `Failed to get token transactions for guildId=${guildId}, userId=${userId}`,
          err instanceof Error ? err : undefined,
        ),
      );
    }
  }

  /**
   * Redeems a redemption code for the user.
   */
  async redeemCode(
    guildId: string,
    userId: string,
    codeStr: string,
  ): Promise<Result<RedemptionResult, DomainError>> {
    return this.redemptionService.redeemCode(codeStr, Number(guildId), Number(userId));
  }

  /**
   * Gets a paginated page of product redemption transactions for a member.
   * Queries the product_redemption_transaction table directly via the DB pool.
   */
  async getProductRedemptionTransactionPage(
    guildId: string,
    userId: string,
    page: number = 1,
    pageSize: number = 10,
  ): Promise<Result<RedemptionTransactionPage, DomainError>> {
    try {
      if (page < 1) page = 1;
      if (pageSize < 1) pageSize = 10;

      // Lazy-resolve the DB pool from the DI container. This avoids coupling
      // this facade to a specific ORM at import time. The pool is expected
      // to be registered at TOKENS.DatabasePool by the shared module.
      const { container, TOKENS } = await import('@ltdjms/shared');

      const db: {
        execute: (query: string, params?: unknown[]) => Promise<Array<Record<string, unknown>>>;
      } = container.resolve(TOKENS.DatabasePool);

      // Query total count — use parameterized query to prevent SQL injection.
      const countSql = 'SELECT COUNT(*) as cnt FROM product_redemption_transaction WHERE guild_id = ? AND user_id = ?';
      const countResult = await db.execute(countSql, [guildId, userId]);
      const totalCount = Number((countResult[0] as Record<string, unknown>)?.cnt ?? 0);
      const totalPages = Math.max(1, Math.ceil(totalCount / pageSize));
      const offset = (page - 1) * pageSize;

      // Query page data — parameterized query.
      const dataSql = 'SELECT * FROM product_redemption_transaction WHERE guild_id = ? AND user_id = ? ORDER BY created_at DESC LIMIT ? OFFSET ?';
      const rows = await db.execute(dataSql, [guildId, userId, pageSize, offset]);

      // TODO(P1-38): Migrate these raw queries to a proper service/repository
      // layer (e.g., RedemptionTransactionRepository) to avoid direct SQL in
      // facades. Parameterized queries mitigate injection risk, but the ideal
      // fix is a dedicated service method on RedemptionService or a new
      // repository class that encapsulates this query logic.

      const items: RedemptionTransactionEntry[] = rows.map(
        (row: Record<string, unknown>) => ({
          id: Number(row.id),
          productName: String(row.product_name ?? ''),
          code: String(row.code ?? ''),
          rewardedAmount: row.rewarded_amount != null ? Number(row.rewarded_amount) : null,
          createdAt: new Date(String(row.created_at)),
        }),
      );

      return new Ok({
        items,
        hasNext: page < totalPages,
        totalPages,
        currentPage: page,
      });
    } catch (err) {
      return new Err(
        DomainError.persistenceFailure(
          `Failed to get redemption transactions for guildId=${guildId}, userId=${userId}`,
          err instanceof Error ? err : undefined,
        ),
      );
    }
  }
}

File: src/main/java/ltdjms/discord/panel/services/MemberInfoFacade.java
package ltdjms.discord.panel.services;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ltdjms.discord.currency.domain.BalanceView;
import ltdjms.discord.currency.services.BalanceService;
import ltdjms.discord.currency.services.CurrencyTransactionService;
import ltdjms.discord.gametoken.services.GameTokenService;
import ltdjms.discord.gametoken.services.GameTokenTransactionService;
import ltdjms.discord.membership.services.MembershipPanelSummary;
import ltdjms.discord.membership.services.MembershipQueryService;
import ltdjms.discord.redemption.services.ProductRedemptionTransactionService;
import ltdjms.discord.redemption.services.RedemptionService;
import ltdjms.discord.shared.DomainError;
import ltdjms.discord.shared.Result;

/**
 * Facade for member information aggregation. Combines currency and game token information for
 * display in user panels.
 */
public class MemberInfoFacade {

  private static final Logger LOG = LoggerFactory.getLogger(MemberInfoFacade.class);

  private final BalanceService balanceService;
  private final GameTokenService gameTokenService;
  private final GameTokenTransactionService gameTokenTransactionService;
  private final CurrencyTransactionService currencyTransactionService;
  private final RedemptionService redemptionService;
  private final ProductRedemptionTransactionService productRedemptionTransactionService;
  private final MembershipQueryService membershipQueryService;

  public MemberInfoFacade(
      BalanceService balanceService,
      GameTokenService gameTokenService,
      GameTokenTransactionService gameTokenTransactionService,
      CurrencyTransactionService currencyTransactionService,
      RedemptionService redemptionService,
      ProductRedemptionTransactionService productRedemptionTransactionService,
      MembershipQueryService membershipQueryService) {
    this.balanceService = balanceService;
    this.gameTokenService = gameTokenService;
    this.gameTokenTransactionService = gameTokenTransactionService;
    this.currencyTransactionService = currencyTransactionService;
    this.redemptionService = redemptionService;
    this.productRedemptionTransactionService = productRedemptionTransactionService;
    this.membershipQueryService = membershipQueryService;
  }

  /**
   * Gets the user panel view for a member in a guild.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @return Result containing UserPanelView on success, or DomainError on failure
   */
  public Result<UserPanelView, DomainError> getUserPanelView(long guildId, long userId) {
    LOG.debug("Getting user panel view for guildId={}, userId={}", guildId, userId);

    Result<BalanceView, DomainError> balanceResult = balanceService.tryGetBalance(guildId, userId);
    if (balanceResult.isErr()) {
      LOG.warn(
          "Failed to get balance for guildId={}, userId={}: {}",
          guildId,
          userId,
          balanceResult.getError().message());
      return Result.err(balanceResult.getError());
    }

    MembershipPanelSummary membershipSummary = getMembershipSummary(userId);
    return buildUserPanelView(guildId, userId, balanceResult.getValue(), membershipSummary);
  }

  /**
   * Gets the user panel view using a precomputed membership summary.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param membershipSummary membership summary shared across guild panels
   * @return Result containing UserPanelView on success, or DomainError on failure
   */
  public Result<UserPanelView, DomainError> getUserPanelView(
      long guildId, long userId, MembershipPanelSummary membershipSummary) {
    LOG.debug(
        "Getting user panel view for guildId={}, userId={} with cached membership summary",
        guildId,
        userId);

    Result<BalanceView, DomainError> balanceResult = balanceService.tryGetBalance(guildId, userId);
    if (balanceResult.isErr()) {
      LOG.warn(
          "Failed to get balance for guildId={}, userId={}: {}",
          guildId,
          userId,
          balanceResult.getError().message());
      return Result.err(balanceResult.getError());
    }

    return buildUserPanelView(guildId, userId, balanceResult.getValue(), membershipSummary);
  }

  private Result<UserPanelView, DomainError> buildUserPanelView(
      long guildId,
      long userId,
      BalanceView balanceView,
      MembershipPanelSummary membershipSummary) {
    long gameTokens = gameTokenService.getBalance(guildId, userId);

    UserPanelView panelView =
        new UserPanelView(
            guildId,
            userId,
            balanceView.balance(),
            balanceView.currencyName(),
            balanceView.currencyIcon(),
            gameTokens,
            membershipSummary);

    LOG.debug(
        "User panel view created: guildId={}, userId={}, currency={}, tokens={}",
        guildId,
        userId,
        balanceView.balance(),
        gameTokens);

    return Result.ok(panelView);
  }

  /**
   * Builds membership tier and period progress for the user panel.
   *
   * @param userId Discord user snowflake
   * @return summary for panel rendering; tier {@link MembershipTier#NONE} when no membership row
   */
  public MembershipPanelSummary getMembershipSummary(long userId) {
    return membershipQueryService.getPanelSummary(userId);
  }

  /**
   * Gets a page of token transaction history for a user.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param page the page number (1-based)
   * @return the transaction page
   */
  public GameTokenTransactionService.TransactionPage getTokenTransactionPage(
      long guildId, long userId, int page) {
    LOG.debug(
        "Getting token transaction page for guildId={}, userId={}, page={}", guildId, userId, page);
    return gameTokenTransactionService.getTransactionPage(
        guildId, userId, page, GameTokenTransactionService.DEFAULT_PAGE_SIZE);
  }

  /**
   * Gets a page of currency transaction history for a user.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param page the page number (1-based)
   * @return the currency transaction page
   */
  public CurrencyTransactionService.TransactionPage getCurrencyTransactionPage(
      long guildId, long userId, int page) {
    LOG.debug(
        "Getting currency transaction page for guildId={}, userId={}, page={}",
        guildId,
        userId,
        page);
    return currencyTransactionService.getTransactionPage(
        guildId, userId, page, CurrencyTransactionService.DEFAULT_PAGE_SIZE);
  }

  /**
   * Redeems a code for a user.
   *
   * @param code the redemption code string
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @return Result containing the redemption result or an error
   */
  public Result<RedemptionService.RedemptionResult, DomainError> redeemCode(
      String code, long guildId, long userId) {
    LOG.debug("Redeeming code for guildId={}, userId={}", guildId, userId);
    return redemptionService.redeemCode(code, guildId, userId);
  }

  /**
   * 取得使用者的商品兌換交易分頁紀錄。
   *
   * @param guildId Discord 伺服器 ID
   * @param userId Discord 使用者 ID
   * @param page 頁碼（從 1 開始）
   * @return 商品兌換交易分頁紀錄
   */
  public ProductRedemptionTransactionService.TransactionPage getProductRedemptionTransactionPage(
      long guildId, long userId, int page) {
    LOG.debug(
        "Getting product redemption transaction page for guildId={}, userId={}, page={}",
        guildId,
        userId,
        page);
    return productRedemptionTransactionService.getTransactionPage(
        guildId, userId, page, ProductRedemptionTransactionService.DEFAULT_PAGE_SIZE);
  }
}
